tracee/event_monitor_ebpf.c

```diff
@@ -1360,188 +1379,222 @@ int trace_ret_##name(struct pt_regs *ctx)                               \
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+    void *ctx = args;
+    regs.di = args->args[0];
+    regs.si = args->args[1];
+    regs.dx = args->args[2];
+    regs.r10 = args->args[3];
+    regs.r8 = args->args[4];
+    regs.r9 = args->args[5];
+    id = args->id;
+#else
+    bpf_probe_read(&regs, sizeof(struct pt_regs), (void*)ctx->args[0]);
+    id = ctx->args[1];
+#endif
+
+    // essential events should always be traced
+    if (id != SYS_CLONE && id != SYS_FORK && id != SYS_VFORK && id != SYS_EXECVE && id != SYS_EXECVEAT) {
+        if (!should_trace())
+            return 0;
+    }
+
+    // We save args here (and not before the tail call function below) as other functions may use it
+    // See mprotect_alert and mmap_alert functions below for comments
+    // If no other function needs the saved args, we can move this to just before the tail call
+    save_args(&regs, id);
+
+    // essential events should always be traced
+    if (id != SYS_CLONE && id != SYS_FORK && id != SYS_VFORK && id != SYS_EXECVE && id != SYS_EXECVEAT) {
+        if (!event_chosen(id) && !event_chosen(RAW_SYS_ENTER))
+            return 0;
+    }
+
+    if (event_chosen(RAW_SYS_ENTER)) {
+        context_t context = {};
+        init_context(&context);
+        set_buf_off(SUBMIT_BUF_IDX, sizeof(context_t));
+        buf_t *submit_p = get_buf(SUBMIT_BUF_IDX);
+        if (submit_p == NULL)
+            return 0;
+
+        context.eventid = RAW_SYS_ENTER;
+        context.argnum = 1;
+        context.retval = 0;
+
+        save_context_to_buf(submit_p, (void*)&context);
+
+        save_to_submit_buf(submit_p, (void*)&id, sizeof(int), INT_T, TAG_SYSCALL);
+        events_perf_submit(ctx);
+    }
+
+    // call syscall handler, if exists
+    sys_enter_tails.call(ctx, id);
+    return 0;
+}
+
+// include/trace/events/syscalls.h:
+// TP_PROTO(struct pt_regs *regs, long ret)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+TRACEPOINT_PROBE(raw_syscalls, sys_exit)
+#else
+RAW_TRACEPOINT_PROBE(sys_exit)
+#endif
+{
+    int id;
+    long ret;
+
+    // we don't support ia32 syscalls at this point
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+    if (task->thread_info.status & TS_COMPAT)
         return 0;
 
-    context.eventid = RAW_SYSCALLS;
-    context.argnum = 1;
-    context.retval = 0;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 17, 0)
+    void *ctx = args;
+    id = args->id;
+    ret = args->ret;
+#else
+    struct pt_regs *regs = (struct pt_regs*)ctx->args[0];
+    id = regs->orig_ax;
+    ret = ctx->args[1];
+#endif
 
-    save_context_to_buf(submit_p, (void*)&context);
+    // essential events should always be traced
+    if (id != SYS_CLONE && id != SYS_FORK && id != SYS_VFORK && id != SYS_EXECVE && id != SYS_EXECVEAT) {
+        if (!should_trace())
+            return 0;
+
+        if (!event_chosen(id) && !event_chosen(RAW_SYS_EXIT))
+            return 0;
+    }
+
+    if (event_chosen(RAW_SYS_EXIT)) {
```

Comment:
why can't we use the generic handler (tail call) here?

Reply:
This is not a syscall - this is the event "sys_exit" which is part of the raw_syscalls, and belongs here.
Anyways, all the tail calls mechanism will change soon to support all syscalls at once (my next PR)